docs/src/index.md

```diff
@@ -27,7 +27,271 @@ this library has a few goals:
+(p1.backtrack | p2 ).parse(" 1")
+// res1: Either[Error, Tuple2[String, Char]] = Right((,1))
+```
+
+Notice that `(p1.backtrack | p2)` clause is another parser by itself since we're still combining parsers by using `orElse`. But we've already used `orElse` in example before without any `backtrack` operator, and it worked just fine. Why do we need `backtrack` now? Let's look at this example to understand where we need to use `backtrack` and where we don't:
+
+```scala
+import cats.parse.Rfc5234.{digit, sp}
+
+val p1 = sp *> digit <* sp
+val p2 = sp *> digit
+val p3 = digit
+
+(p1 | p2).parse(" 1")
+// res1 = Left(Error(2,NonEmptyList(InRange(2, , ))))
+
+(p1 | p2 | p3).parse("1")
+// res2 = Right((,1))
+```
+
+The first parsing is interrupted by error since it *started* the parsing (one first space was correct on parser `p1`) so it won't work without `backtrack`. The second parsing works fine with `orElse` alone since it does not start in `p1` or `p2`.
+
+So the `backtrack` helps us where the *left side* throws an error.
+
+### Soft
+
+This method might look similar to `backtrack`, but it allows us to *proceed* the parsing when the *right side* is failed. It is really useful for ambiguous parsers when we can't really tell what exactly we are parsing before the end. Let's say we want to parse some input to the search engine which contains fields. This might look like "field:search_query". Let's try to write a parser for this:
+
+```scala
+import cats.parse.Rfc5234.{alpha, sp}
+import cats.parse.Parser
+import cats.parse.Parser.{char => pchar}
+
+val searchWord = alpha.rep.string
+
+val fieldValue = alpha.rep.string ~ pchar(':')
+
+val p1 = fieldValue.? ~ (searchWord ~ sp.?).rep.string
+
+
+p1.parse("title:The Wind Has Risen")
+// res0 = Right((,(Some((title,())),The Wind Has Risen)))
+p1.parse("The Wind Has Risen")
+// res1 = Left(Error(3,NonEmptyList(InRange(3,:,:))))
+```
+
+This error happens because we can't really tell if we are parsing the `fieldValue` before we met a `:` char. We might do this with `backtrack` only by writing two parsers:
+
+```scala
+val p1 = fieldValue.? ~ (searchWord ~ sp.?).rep.string
+
+val p3 = (searchWord ~ sp.?).rep.string
+
+(p1.backtrack | p3).parse("title:The Wind Has Risen")
+// res0 = Right((,(Some((title,())),The Wind Has Risen)))
+(p1.backtrack | p3).parse("The Wind Has Risen")
+// res1 = Right((,The Wind Has Risen))
+```
+
+But this problem might be resolved with `soft` method:
+
+```scala
+val searchWord = alpha.rep.string
+
+val fieldValueSoft = alpha.rep.string.soft ~ pchar(':')
+
+val p2 = fieldValueSoft.? ~ (searchWord ~ sp.?).rep.string
+
+p2.parse("title:The Wind Has Risen")
+// res2 = Right((,(Some((title,())),The Wind Has Risen)))
+p2.parse("The Wind Has Risen")
+// res3 = Right((,(None,The Wind Has Risen)))
+```
+
+So when the *right side* throws an error the `soft` method allows us to rewind parsed input and try to proceed it's parsing with next parsers (without changing the parser itself!).
```

Comment:
when the right side throws a *non-halting* error (which is to say epsilon error, or the right side did not consume any input).